crates/runtime/src/datafusion/spice_udf_override.rs

```diff
@@ -0,0 +1,290 @@
+#[derive(Default, Debug)]
+pub struct SpiceUDFsOverride {}
+
+impl SpiceUDFsOverride {
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl OptimizerRule for SpiceUDFsOverride {
+    fn apply_order(&self) -> Option<ApplyOrder> {
+        Some(ApplyOrder::TopDown)
+    }
+
+    fn supports_rewrite(&self) -> bool {
+        true
+    }
+
+    fn rewrite(
+        &self,
+        plan: LogicalPlan,
+        _config: &dyn OptimizerConfig,
+    ) -> Result<Transformed<LogicalPlan>, DataFusionError> {
+        if let LogicalPlan::Extension(extension_node) = plan {
+            if let Some(node) = extension_node
+                .node
+                .as_any()
+                .downcast_ref::<FederatedPlanNode>()
+            {
+                // extension plan node and logical plan are immutable references
+                // so we need to clone the plan and reconstruct the extension node
+                let updated_plan = node.plan().clone().transform(override_spice_udf)?;
+                let planner = node.planner();
+
+                let updated_extension_node = Extension {
+                    node: Arc::new(FederatedPlanNode::new(updated_plan.data, planner)),
+                };
+
+                if updated_plan.transformed {
+                    return Ok(Transformed::yes(LogicalPlan::Extension(updated_extension_node)));
+                }
+                return Ok(Transformed::no(LogicalPlan::Extension(updated_extension_node)));
+            }
+            return Ok(Transformed::no(LogicalPlan::Extension(extension_node)));
+        }
+
+        Ok(Transformed::no(plan))
+    }
+
+    fn name(&self) -> &str {
+        "spiceai_udf_override_rule"
+    }
+}
+
+/// Replaces `SpiceAI` UDFs with target execution engine-specific UDFs within the logical plan.
+/// The transformation applies only to UDFs that are part of `Projection` nodes.
+fn override_spice_udf(plan: LogicalPlan) -> DataFusionResult<Transformed<LogicalPlan>> {
+    match plan {
+        LogicalPlan::Projection(Projection { expr, input, .. }) => {
+            // Note: while there is currently no in-place mutation API that uses `&mut TreeNode`,
+            // the transforming APIs are efficient and optimized to avoid cloning.
+            match override_spice_udf_in_exprs(&input, expr)? {
+                (new_expr, true) => Ok(Transformed::yes(
+                    Projection::try_new(new_expr, Arc::clone(&input))
+                        .map(LogicalPlan::Projection)?,
+                )),
+                (new_expr, false) => Ok(Transformed::no(
+                    Projection::try_new(new_expr, Arc::clone(&input))
+                        .map(LogicalPlan::Projection)?,
+                )),
+            }
+        }
+        _ => Ok(Transformed::no(plan)),
+    }
+}
+
+fn override_spice_udf_in_exprs(
+    input: &LogicalPlan,
+    exprs: Vec<Expr>,
+) -> DataFusionResult<(Vec<Expr>, bool)> {
```

Comment:
Instead of returning a bool to indicate modified, could we reuse the Transformed::yes/no here?